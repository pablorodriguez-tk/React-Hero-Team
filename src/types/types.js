export const types = {
  authLogin: '[auth] Login',
  authLogout: '[auth] Logout',

  heroSearch: '[hero] Search',
  heroAdd: '[Hero] Add to team',
  HeroLogout: '[hero] Logout',
  HeroGetById: '[hero] Get by ID',
  HeroId: '[Hero] Id hero saved',
  HeroFetch: '[Hero] hero Fetch',

  validationBadHero: '[validation] add bad hero',
  validationGoodHero: '[validation] Add good hero',
  validationNeutralHero: '[validation] Add neutral hero',
  validationLogout: '[validation] Logout',
};
