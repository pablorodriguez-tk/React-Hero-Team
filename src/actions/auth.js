import { loginWithEmailAndPassword } from '../helpers/fetch';
import { types } from '../types/types';

export const startLogin = ({ email, password }) => {
  return async (dispatch) => {
    const { data } = await loginWithEmailAndPassword(email, password);
    const { token } = data;
    if (token) {
      localStorage.setItem('token', token);
      dispatch(login(email));
    }
  };
};

export const login = (email) => ({
  type: types.authLogin,
  payload: {
    email,
  },
});

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
  };
};

export const logout = () => ({
  type: types.authLogout,
});
