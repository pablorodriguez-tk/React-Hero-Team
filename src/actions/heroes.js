import { searchHero, getHeroById } from '../helpers/fetch';
import { types } from '../types/types';

export const startSearch = (query) => {
  return async (dispatch) => {
    const data = await searchHero(query);
    const heroes = data;

    if (heroes) {
      dispatch(search(heroes));
    }
  };
};

const search = (heroSearch) => ({
  type: types.heroSearch,
  payload: heroSearch,
});

export const HeroAdd = ({
  id,
  name,
  powerstats,
  biography,
  appearance,
  work,
  connections,
  image,
}) => ({
  type: types.heroAdd,
  payload: {
    id,
    name,
    powerstats,
    biography,
    appearance,
    work,
    connections,
    image,
  },
});

export const HeroLogout = () => ({
  type: types.HeroLogout,
});

export const startGetHeroesById = (id) => {
  return async (dispatch) => {
    try {
      const data = await getHeroById(id);
      return data;
    } catch (error) {
      console.log(error);
    }
  };
};

export const HeroId = (id) => ({
  type: types.HeroId,
  payload: id,
});

export const heroFetch = (data) => ({
  type: types.HeroFetch,
  payload: data,
});
