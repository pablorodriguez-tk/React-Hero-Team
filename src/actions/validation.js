import { types } from '../types/types';

export const AddedBadHero = () => ({
  type: types.validationBadHero,
});

export const AddedGoodHero = () => ({
  type: types.validationGoodHero,
});

export const AddedNeutralHero = () => ({
  type: types.validationNeutralHero,
});

export const validationLogout = () => ({
  type: types.validationLogout,
});
