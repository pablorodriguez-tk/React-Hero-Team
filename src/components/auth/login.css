.login-container {
  margin-top: 15%;
  max-width: 60%;
}
.login-form {
  border-radius: 5px;
  box-shadow: 0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 9px 26px 0 rgba(0, 0, 0, 0.19);
  padding: 5%;
}
.login-form h3 {
  text-align: center;
  color: #333;
}
.login-container form {
  padding: 5%;
}
.btnSubmit {
  width: 50%;
  border-radius: 1rem;
  padding: 1.5%;
  border: none;
  cursor: pointer;
}
.login-form .btnSubmit {
  color: #fff;
  background-color: #0062cc;
}

.login-form .btnSubmit:hover {
  color: #fff;
  background-color: #014491;
}
