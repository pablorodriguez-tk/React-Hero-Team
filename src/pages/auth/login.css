.login-container {
  margin-top: 10%;
  max-width: 80%;
}

.login-container-search {
  margin-top: 2%;
  max-width: 60%;
}

.login-form {
  border-radius: 5px;
  box-shadow: 0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 9px 26px 0 rgba(0, 0, 0, 0.19);
  padding: 5%;
}
.login-form h3 {
  text-align: center;
  color: #333;
}
.login-container form {
  padding: 5%;
}
.btnSubmit {
  width: 50%;
  border-radius: 1rem;
  padding: 1.5%;
  border: none;
  cursor: pointer;
}
.login-form .btnSubmit {
  color: #fff;
  background-color: #0062cc;
}

.login-form .btnSubmit:hover {
  color: #fff;
  background-color: #014491;
}

.login-form .btnSubmit:disabled {
  color: #fff;
  background-color: grey;
}

.error {
  position: absolute;
  color: red;
  font-size: 0.7rem;
}
