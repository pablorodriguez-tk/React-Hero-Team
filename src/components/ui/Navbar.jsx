import React from 'react';
import { useDispatch } from 'react-redux';
import { Link, NavLink } from 'react-router-dom';
import { startLogout } from '../../actions/auth';

export const Navbar = () => {
  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <div className="container">
        <Link className="navbar-brand" to="/heroteam">
          Hero Team
        </Link>

        <div className="navbar-collapse">
          <div className="navbar-nav">
            <NavLink
              activeClassName="active"
              className="nav-item nav-link"
              exact
              to="/search"
            >
              Search
            </NavLink>
          </div>
        </div>

        <ul className="navbar-nav ml-auto">
          <button className="nav-item nav-link btn" onClick={handleLogout}>
            Logout
          </button>
        </ul>
      </div>
    </nav>
  );
};
