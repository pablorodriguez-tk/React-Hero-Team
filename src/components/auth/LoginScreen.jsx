import React from 'react';
import { useDispatch } from 'react-redux';
import { startLogin } from '../../actions/auth';
import './login.css';
import { Formik } from 'formik';
import * as Yup from 'yup';

export const LoginScreen = () => {
  const dispatch = useDispatch();

  const handleSubmit = async (values, { setSubmitting }) => {
    setSubmitting(false);
    await dispatch(startLogin(values));
  };

  return (
    <Formik
      //TODO: In production, this initial form must be a empty string
      initialValues={{ email: 'challenge@alkemy.org', password: 'react' }}
      validationSchema={Yup.object({
        email: Yup.string().email('Invalid email address').required('Required'),
        password: Yup.string()
          .min(5, 'Password must be at least 5 characters')
          .required('Required'),
      })}
      onSubmit={handleSubmit}
    >
      {(formik) => (
        <div className="container login-container animate__animated animate__fadeIn animate__faster ">
          <div className="col-md-12 login-form">
            <h3 data-testid="loginScreen">Login</h3>
            <form onSubmit={formik.handleSubmit}>
              <div className="input-group mb-1">
                <input
                  placeholder="Email"
                  className={`form-control ${
                    formik.touched.email && formik.errors.email && 'is-invalid'
                  }`}
                  id="email"
                  type="email"
                  autoComplete="off"
                  {...formik.getFieldProps('email')}
                />
              </div>

              <div className="input-group mb-4">
                {formik.touched.email && formik.errors.email ? (
                  <div className="error" data-testid="email-error">
                    {formik.errors.email}
                  </div>
                ) : null}
              </div>

              <div className="input-group mb-1">
                <input
                  placeholder="Password"
                  className={`form-control ${
                    formik.touched.password &&
                    formik.errors.password &&
                    'is-invalid'
                  }`}
                  id="password"
                  type="password"
                  {...formik.getFieldProps('password')}
                />
              </div>

              <div className="input-group mb-5">
                {formik.touched.password && formik.errors.password ? (
                  <div className="error" data-testid="password-error">
                    {formik.errors.password}
                  </div>
                ) : null}
              </div>

              <button
                className="btnSubmit mb-3"
                type="submit"
                disabled={!formik.isValid}
              >
                Login
              </button>
            </form>
          </div>
        </div>
      )}
    </Formik>
  );
};
