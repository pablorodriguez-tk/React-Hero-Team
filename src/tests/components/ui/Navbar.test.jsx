import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router';
import { Navbar } from '../../../components/ui/Navbar';
import { startLogout } from '../../../actions/auth';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const initState = {
  auth: {
    isLoggedIn: true,
    email: 'challenge@alkemy.org',
  },
  heroes: {
    heroTeam: [],
  },
  _persist: {
    version: -1,
    rehydrated: true,
  },
};

let store = mockStore(initState);

store.dispatch = jest.fn();

describe('Test on <Navbar/>', () => {
  beforeEach(() => jest.clearAllMocks());

  test('Should be displayed correctly', () => {
    const wrapper = render(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
    const logout = wrapper.queryByText('Logout');
    expect(logout).toBeInTheDocument();
  });

  test('Should call logout and use history', () => {
    const initState = {};
    store = mockStore(initState);
    const wrapper = render(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>
    );

    // To see problem
    const expectedActions = [
      {
        type: types.authLogout,
      },
      {
        type: types.HeroLogout,
      },
      {
        type: types.authLogout,
      },
      {
        type: types.HeroLogout,
      },
    ];

    fireEvent.click(wrapper.getByTestId('logoutButton'));
    store.dispatch(startLogout());
    expect(store.getActions()).toEqual(expectedActions);
  });
});
