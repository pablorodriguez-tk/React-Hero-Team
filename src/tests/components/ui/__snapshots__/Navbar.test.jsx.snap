// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <Navbar/> Should be displayed correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-sm navbar-dark bg-dark"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/heroteam"
          >
            Hero Team
          </a>
          <div
            class="navbar-collapse"
          >
            <div
              class="navbar-nav"
            >
              <a
                class="nav-item nav-link"
                href="/search"
              >
                Search
              </a>
            </div>
          </div>
          <ul
            class="navbar-nav ml-auto"
          >
            <button
              class="nav-item nav-link btn"
              data-testid="logoutButton"
            >
              Logout
            </button>
          </ul>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-sm navbar-dark bg-dark"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand"
          href="/heroteam"
        >
          Hero Team
        </a>
        <div
          class="navbar-collapse"
        >
          <div
            class="navbar-nav"
          >
            <a
              class="nav-item nav-link"
              href="/search"
            >
              Search
            </a>
          </div>
        </div>
        <ul
          class="navbar-nav ml-auto"
        >
          <button
            class="nav-item nav-link btn"
            data-testid="logoutButton"
          >
            Logout
          </button>
        </ul>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
